Bitmap2level aka png2obj

iacopy opened the pull request:
Implement #13 

- [x] good coverage
- [x] mypy is happy
- [x] flake8 is happy
- [x] speed is OK
- [x] docstrings
- [x] adjust import paths
- [x] adjust naming and phrasing (node, vertex, box, etc)
- [x] Remove blender plugin
- [x] Remove complex doctests
- [ ] zip (test) mesh data

To move in another branch / PR
- [x] Blender plugin freeze issue
- [x] Floor

Commit:
bmp2obj - simple floor obj creator: each pixel becomes a separated square in the obj

Commit:
WIP things

Commit:
Add downloaded shapedetector as is

Commit:
Style: pep8ify shape detector

Commit:
wait 1 s for each shape

Commit:
Detect the minimal points to describe the geometry; rename ugly variables in meaningful ways

Commit:
Remove the shape detector dependency

Commit:
Try with Ivan's rdp (Ramer-Douglas-Peucker) implementation

Commit:
Revert "Try with Ivan's rdp (Ramer-Douglas-Peucker) implementation"

This reverts commit e371f5f022ed5abfced7d9a8f133099f7b9c09e5.

Commit:
add image sample that shows the failing behaviour of the algorithm

Commit:
mv bmp2obj.py

Commit:
bitmap2level: Add test for the edge detector

Commit:
wip edge finder (pass 4/15 tests); do not remove internal edge vertices

Commit:
edge finder: remove internal edge points: 4/15 -> 11/15 tests pass now

Commit:
edge finder: rename class

Commit:
edge finder: fixup function type hints

Commit:
edge detector: fixup doctest

Commit:
edge detector: cleanup old unused code

Commit:
edge detector: print become logging.debug

Commit:
edge detector: cleanup commented out code and prints

Commit:
update.gitignore adding 'htmlcov' directory

Commit:
edge detector: cleanup - remove failed unused files

Commit:
edge detector: cleanup unused functions

Commit:
edge detector: add 2 test cases for single shapes ("snake" and "H")

Commit:
png2obj: add 2 modules for edge extrusion and wavefront export

wip: not 100% completed but 100% tested (doctest)

Commit:
Add pytest and coverage config files

Commit:
png2obj: type hints for the package

Commit:
edge detector: minor changes (relocate test, add a simple one, comments)

Commit:
Cleanup

Commit:
edge detector: small refactoring (to prepare support multiple walls)

Commit:
edge detector: add more test cases (variations for o-room)

Commit:
ignore mesh_converter.py and demo.py in pytest

Commit:
edge detector: implement multiple lines support

Before, just a single connected line was detected,
so, with a single non-closed wall it was ok, but
for a closed room, only the external wall surface
was detected, for example.

Commit:
png2obj: add loader png2matrix with tests. Given a png it returns a walkable matrix

Commit:
png2obj/cleanup: removed unused sample map

Commit:
edge detector: add hint to fix mypy

Commit:
edge detector: cleanup unused variable

Commit:
png2obj: improve the png loader (mode 'P' png image, like 1 byte for each pixel)

Commit:
png2obj: create main function `png2obj` and generic test cases for it

Commit:
png2obj: no relative imports

Commit:
png2onj: first optimization (class removed) 2.5x faster

Commit:
png2obj: print main steps in main function

Commit:
png2obj: improve speed of edge detection by using a map for tracked_vertices

Commit:
png2obj: faster (> 50x) obj exporting

Commit:
png2obj: faster remove_internal_edge_points

Commit:
png2obj: faster edge detection (avoid to create numpy arrays to sum 2 vertex)

Commit:
png2obj: cleanup (comments, flake8, etc)

Commit:
png2obj: 27x faster by removing `tracked` list

Commit:
png2obj: remove redundant check (small speed improvement)

Commit:
png2obj: add support for axes export option (ex. -Z forward, Y Up, ...)

Commit:
png2obj: docstrings

Commit:
png2obj: file encoding

Commit:
png2obj: improve a type hint

Commit:
png2obj: small refactor of 'main' function

Commit:
remove 'source' option from .coveragerc

Commit:
png2obj: use argparse

arl commented on config/game.ini:
```diff
@@ -4,8 +4,8 @@ ServerPort = 1234
-Width = 1024
-Height = 768
+Width = 2560
+Height = 1440
```
Does this should be commited?

iacopy replied:
No no good catch, it was committed by mistake!

arl commented on requirements.txt:
```diff
@@ -1,10 +1,12 @@
 line-profiler==1.0
 mccabe==0.4.0
 msgpack-python==0.4.7
+mypy-lang===0.4.6-dev-1d2a046c149e010816b6bcd5feae49e0ce3b32d2
```
I'm not sure we should add it to the requirements. Does pip has the concept of `devDependencies` like nodeJS? I mean, `mypy` we shoulld try to keep things splitted between what is needed to:
 - build the game
 - develop the game
 - run the game
 
This to not further add to the complexity of continuous integration and the creation of a python binary. Overall i say that because png2obj is an external tool at destination of content creators, maybe it should have its own `requirements.txt`

arl commented on tools/png2obj/png2obj.py:
```diff
@@ -0,0 +1,362 @@
+}  # type: Dict[Vector2D, str]
+
+
+# A vertex ha 4 neighbour boxes, and each box can be walkable or not (block).
```
typo ha -> has

Commit:
Add tests for the "chessboard" case (currently failing)

Commit:
Adding debug info

Commit:
fixup chessboard tests

Commit:
test: add other 2 chessboard cases

Commit:
png2obj: refactor and fix chessboard case (TODO: remove debugging stuff)

Commit:
png2ob: remove functions/methods not used anymore

Commit:
remove debug prints but keep turtle debug optionally

Commit:
Add samples

Commit:
png2obj: try to triangulate top horizontal walls surfaces

works with H shape, but fails with 0 and other ones

Commit:
Update requirements for triangle library (itself requires Cython and Numpy)

Commit:
fixup create_wavefront doctest

Commit:
change DEFAULT_EXPORT_SETTINGS to ('x', 'y', 'z') to have z up instead of down in Blender

Commit:
update 4x4_chessboard_1.obj after export settings update (x, y, z)

Commit:
update 4x4_chessboard_1.obj after top horizontal surface wall triangulation

Commit:
png2obj: triangulate multiple walls

This is affected by the triangle.c segmentation fault in some cases
since we pass 1 hole per each walkable cell

Commit:
png2obj: fix turtle cli argument

Commit:
png2obj: refactor - separate 'matrix2obj' task into a new function

Thought to be used by stress tests

Commit:
png2obj: reduce the risk to have the triangulate segmentation fault by
passing not all walkable cells as holes but only the ones which surround
the walls.

Commit:
png2obj: use red pixels to set -1 in the walkable matrix

Commit:
png2obj: create a costant for the turtle

Commit:
png2obj: turtle speed from 9 to 11

Commit:
img2gif: remove print(mesh)

Commit:
png2obj: rename 'debug' flag to 'turtle' (and also fix passing that arg to matrix2obj)

Commit:
png2obj: (m) save triangle input in json instead of pickle and update
module docstring

Commit:
coverage configuration: omit .virtualenvs

Commit:
add tests for rect tessellation

Commit:
fixup

Commit:
skip test_cover since the tessellation will be delegated to Blender

Commit:
Refactor png2obj: replace BlocksMap class with a function (there is no state to handle...)

Commit:
Refactor png2obj: remove redundant dict POSSIBLE_DIRECTIONS and use RULES

Commit:
Reorder imports

Commit:
Refactor png2obj: rename box* to cell*

Commit:
png2obj: drop triangulation stuff (-> Blender)

Commit:
png2obj: rename variables to avoid conflict when merging modules for blender plugin (bovin)

Commit:
png2obj: blender add-on WIP

Commit:
png2obj: wall_perimeters_to_unique_vertices -> wall_perimeters_to_verts_edges

Currently unused function, will be used in the blender plugin

Commit:
Bleveler: make first working blender-addon version (to wrap-up and clean-up)

Commit:
Bleveler: fix import positions, style and prints.

Commit:
Bleveler: Cleanup dead functions (load_png, matrix2obj, png2obj)

Commit:
bleveler: Fix geometric issue due to wrong image loading.

Commit:
bleveler: fix doc

Commit:
bleveler: refactoring: remove unused gui widgets (button and text field) fix and change names/phrasing etc

Commit:
Bleveler: small image loading improvement

Commit:
Bleveler: add debug prints (measure steps time)

Commit:
Bleveler: make wall height settable by GUI

Commit:
fixup

Commit:
Bleveler: make de cell (pixel) size settable by GUI

Commit:
Bleveler: rename cell_size to pixel_size

Commit:
Bleveler: improve documentation

Commit:
Bleveler: set the toolbar category (tab name) to "Surviveler" (was "Misc" by default)

Commit:
Bleveler: floor. (closes #13)

iacopy commented:
Floor added.

voidexp commented on tools/png2obj/tests/samples/4x4_chessboard_1.obj:
```diff
@@ -0,0 +1,95 @@
+v 0.000000 0.000000 1.000000
```
It's relatively easy to read compressed files via `gzip` module. Could we gzip this one, so it doesn't figure out as plain text?

iacopy replied:
Yes of course

voidexp commented on tools/png2obj/bleveler.py:
```diff
@@ -0,0 +1,560 @@
+# -*- coding: utf-8 -*-
```
This one duplicates a lot of code which is part of normal CLI tool. Maintaining one would imply to go around the other and apply changes. If there's no easy way so far to create a Blender plugin with importable external modules, lets leave this out for now.

voidexp commented:
Personally, I don't like the idea of example code in comments. A short docstring + typings is more than enough to understand what a function does. If it is not, maybe the function should be split or refactored to be simpler or have a more meaningful name. Taking the example at https://github.com/RookieGameDevs/surviveler/pull/5/files#diff-eb8eea8ea9427b5f05f9b0edc99fa94cR164 ... It would be much better effort to document the function and it's parameters, instead of providing usage example. Another big drawback of this approach: somebody has to maintain examples... If function prototype/name/number of arguments change, one has to recreate the example, copy the results in the docstring, ensure they're correct, etc. Also, this would make diffs cluttered with useless stuff.

iacopy commented:
@V0idExp Actually those examples are **tests** (`doctests`), executed when you run `py.test`.
Personally, in general, I often strongly prefer *live* examples than words (if I have to choose, but much better to have both), at least for simple functions. The maintenance is exactly the same than the one you have to do with tests. The advantage is that you immediately see what the function does. Anyway, every kind of documentation needs to be maintained then the code changes. But this way, you know when it has to be updated since the tests fail.

voidexp commented:
As said by voice, things which should be tests, should be turned into tests. And tests *must* be maintained, thus, if one wants to look for usage examples, it's a matter of just opening a test suite and look into. And docstrings are short, clear and possibly, with documented parameters.

voidexp commented:
For the Blender plugin, let's remove it from this PR, and figure out a way to base a plugin on the current code that could be importable and not a monolithic piece of Python code, which is also duplicated. For the rest, very good work, clear, modularized and well readable. Didn't check algorithmic part, that would be too much for my brain after a whole day programming :)

iacopy committed:
Move pytest.ini inside the png2obj directory.

iacopy committed:
Remove blender plugin (to be added later).

iacopy committed:
Fix imports.

iacopy committed:
Cleanup: remove useless obj.

iacopy committed:
png2obj: docstring - move long doctests to new test_wavefront.py module and add few docstrings

iacopy committed:
png2obg: remove triangle stuff (unused since we plan to use blender to make tessellation)

iacopy committed:
png2obj: add a simple readme.

iacopy committed:
png2obj: rename walkable matrix to walkability matrix

iacopy committed:
fixup config/game.ini

iacopy committed:
fix mypy requirements.txt (no specific version)

iacopy committed:
png2obj: remove triangle requirement.

iacopy committed:
png2obj: add tests to Travis CI
